tools/resource-gen/main.go

```diff
@@ -418,34 +405,99 @@ func main() {
 		resources = append(resources, resourceInfo)
 	}
 
-	var generatorTemplate *template.Template
+	var generatorFn GeneratorFn
 
 	switch gen {
 	case "type":
-		generatorTemplate = ResourceTemplate
+		generatorFn = TemplateGeneratorFn(ResourceTemplate)
 	case "crd":
-		generatorTemplate = CustomResourceTemplate
+		generatorFn = TemplateGeneratorFn(CustomResourceTemplate)
+	case "openapi":
+		generatorFn = openApiGenerator
 	default:
 		log.Fatalf("%s is not a valid generator option\n", gen)
 	}
 
-	outBuf := bytes.Buffer{}
-	if err := generatorTemplate.Execute(&outBuf, struct {
-		Package   string
-		Resources []ResourceInfo
-	}{
-		Package:   pkg,
-		Resources: resources,
-	}); err != nil {
-		log.Fatalf("template error: %s", err)
-	}
-
-	out, err := format.Source(outBuf.Bytes())
+	err := generatorFn(pkg, resources)
 	if err != nil {
 		log.Fatalf("%s\n", err)
 	}
+}
 
-	if _, err := os.Stdout.Write(out); err != nil {
-		log.Fatalf("%s\n", err)
+func openApiGenerator(pkg string, resources []ResourceInfo) error {
+	protoTypeToType := map[string]reflect.Type{
+		"Mesh":        reflect.TypeOf(v1alpha1.Mesh{}),
+		"MeshGateway": reflect.TypeOf(v1alpha1.MeshGateway{}),
 	}
+
+	for _, r := range resources {
+		tpe, exists := protoTypeToType[r.ResourceType]
+		println("generating", r.ResourceType)
+		if !exists {
+			continue
+		}
+		reflector := jsonschema.Reflector{
+			ExpandedStruct:            true,
+			DoNotReference:            true,
+			AllowAdditionalProperties: true,
+		}
+		err := reflector.AddGoComments("github.com/kumahq/kuma/", "api/")
+		if err != nil {
+			return err
+		}
+		schemaMap := orderedmap.New[string, *jsonschema.Schema]()
+		schemaMap.Set("type", &jsonschema.Schema{Type: "string"})
+		schemaMap.Set("name", &jsonschema.Schema{Type: "string"})
+		if r.Global {
+			schemaMap.Set("mesh", &jsonschema.Schema{Type: "string"})
+		}
+		schemaMap.Set("labels", &jsonschema.Schema{Type: "object", AdditionalProperties: &jsonschema.Schema{Type: "string"}})
+		schemaMap.Set("spec", reflector.ReflectFromType(tpe))
+
+		schema := jsonschema.Schema{
+			Type:       "object",
+			Required:   []string{"type", "name", "spec"},
+			Properties: schemaMap,
+		}
+
+		// TODO merge the schema with a global open api schema for the resource (might be worth to just make these objects
```

Comment:
merging is done by openapi-tool written in node-js so not sure if anything is needed here, I'll do this in the text commit

Reply:
done merging the way I described above